File: others/WorksApplication/Worksap Package/workspace/China_Intern/China_Intern_Java/src/jp/co/worksap/intern/util/Utilities.java
package jp.co.worksap.intern.util;

import java.io.FileReader;
import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import jp.co.worksap.intern.constants.Messages;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

import au.com.bytecode.opencsv.CSVReader;
import au.com.bytecode.opencsv.CSVWriter;

/**
 * Miscellaneous utilities.
 * 
 */
public final class Utilities {

	private static final Log logger = LogFactory.getLog(Utilities.class);

	// Must not be instantiated
	private Utilities() {
	}

	/**
	 * close reader without any exception.
	 * 
	 * @param reader
	 */
	public static void closeSilently(FileReader reader) {
		if (reader != null) {
			try {
				reader.close();
			} catch (IOException e) {
				logger.fatal(Messages.ERROR_CSV_FILE_CLOSE, e);
			}
		}
	}

	/**
	 * close reader without any exception.
	 * 
	 * @param reader
	 */
	public static void closeSilently(CSVReader reader) {
		if (reader != null) {
			try {
				reader.close();
			} catch (IOException e) {
				logger.fatal(Messages.ERROR_CSV_FILE_CLOSE, e);
			}
		}
	}

	/**
	 * flush first, and close writer without any exception.
	 * 
	 * @param writer
	 */
	public static void closeSilently(CSVWriter writer) {
		if (writer != null) {
			try {
				writer.flush();
				writer.close();
			} catch (IOException e) {
				logger.fatal(Messages.ERROR_CSV_FILE_CLOSE, e);
			}
		}
	}

	/** default datetime format */
	private static final SimpleDateFormat datetimeFormat = new SimpleDateFormat(
			"yyyy/MM/dd HH:mm");

	public static synchronized Date parseDateTimeStr(String datetimeStr)
			throws IllegalArgumentException {
		try {
			return datetimeFormat.parse(datetimeStr);
		} catch (ParseException e) {
			logger.fatal(Messages.ERROR_PARSE_DATE, e);
			throw new IllegalArgumentException("Failed to parse Date Time String!");
		}
	}
	
	public static synchronized String formatDateTime(Date date) {
		return datetimeFormat.format(date);
	}
	
	/** default date format */
	private static final SimpleDateFormat dateFormat = new SimpleDateFormat(
			"yyyy/MM/dd");

	public static synchronized Date parseDateStr(String dateStr)
			throws IllegalArgumentException {
		try {
			return dateFormat.parse(dateStr);
		} catch (ParseException e) {
			logger.fatal(Messages.ERROR_PARSE_DATE, e);
			throw new IllegalArgumentException("Failed to parse Date String!");
		}
	}
	
	public static synchronized String formatDate(Date date) {
		return dateFormat.format(date);
	}

	/** default time format */
	public static final SimpleDateFormat timeFormat = new SimpleDateFormat(
			"HH:mm");

	public static synchronized Date parseTimeStr(String timeStr)
			throws IllegalArgumentException {
		try {
			return timeFormat.parse(timeStr);
		} catch (ParseException e) {
			logger.fatal(Messages.ERROR_PARSE_DATE, e);
			throw new IllegalArgumentException("Failed to parse Time String!");
		}
	}

	public static synchronized String formatTime(Date time) {
		return timeFormat.format(time);
	}
}

File: others/WorksApplication/Worksap Package/workspace/China_Intern/China_Intern_Cpp/src/util/utilities.cpp
#include "utilities.h"
#include "../csv_parser/csv_reader.h"
#include "../csv_parser/csv_writer.h"
#include "strptime.h"
#include <iostream>
#include <ctime>
#include <stdexcept>
using namespace std;

const string Utilities::kDefaultDateWithoutTimeFormat = "%Y/%m/%d";
const string Utilities::kDefaultDateFormat = "%Y/%m/%d %H:%M";
const string Utilities::kDefaultTimeFormat = "%H:%M";

void Utilities::close_siently(ifstream &in) {
	if (in && in.is_open()) {
		in.close();
	}
}

void Utilities::close_siently(CSVReader &reader) {
	reader.close();
}

void Utilities::close_siently(CSVWriter &writer) {
	writer.close();
}

struct tm Utilities::parse_date_without_time_str(string &date_str) {
	struct tm time_struct;
	if (!strptime(date_str.c_str(), Utilities::kDefaultDateWithoutTimeFormat.c_str(),
			&time_struct)) {
		throw invalid_argument("Failed to parse Date String! Please use format: " + Utilities::kDefaultDateWithoutTimeFormat);
	}
	return time_struct;
}

string Utilities::format_date_without_time(struct tm &date) {
	char buf[256] = {0};
	strftime(buf, 256, Utilities::kDefaultDateWithoutTimeFormat.c_str(), &date);
	string date_str(buf);
	return date_str;
}

struct tm Utilities::parse_date_str(string &date_str) {
	struct tm time_struct;
	if (!strptime(date_str.c_str(), Utilities::kDefaultDateFormat.c_str(),
			&time_struct)) {
		throw invalid_argument("Failed to parse Date String! Please use format: " + Utilities::kDefaultDateFormat);
	}
	return time_struct;
}

string Utilities::format_date(struct tm &date) {
	char buf[256] = {0};
	strftime(buf, 256, Utilities::kDefaultDateFormat.c_str(), &date);
	string date_str(buf);
	return date_str;
}

struct tm Utilities::parse_time_str(string &time_str) {
	struct tm time_struct;
	if (!strptime(time_str.c_str(), Utilities::kDefaultTimeFormat.c_str(),
			&time_struct)) {
		throw invalid_argument("Failed to parse Time String! Please use format: " + Utilities::kDefaultTimeFormat);
	}
	return time_struct;
}

string Utilities::format_time(struct tm &time) {
	char buf[256] = {0};
	strftime(buf, 256, Utilities::kDefaultTimeFormat.c_str(), &time);
	string time_str(buf);
	return time_str;
}

